jupyterhub/spawner.py

```diff
@@ -797,8 +797,27 @@ def get_env(self):
             'activity',
         )
         env['JUPYTERHUB_BASE_URL'] = self.hub.base_url[:-4]
+
         if self.server:
+            base_url = self.server.base_url
+            if self.ip or self.port:
+                self.server.ip = self.ip
+                self.server.port = self.port
             env['JUPYTERHUB_SERVICE_PREFIX'] = self.server.base_url
+        else:
+            # this should only occur in mock/testing schenarios
+            base_url = '/'
+
+        if self.ip or self.port:
+            # specify JUPYTERHUB_SERVICE_URL *if* ip or port is specified
+            # TODO: set this always?
+            # Prior to 2.0, leaving 'ip' or 'port' unset meant letting the subprocess default be used
+            # setting via a URL means we cannot specify explicit ip or port without specifying *both*
+            # this results in a changed default behavior of specifying only report
+            # specifying ip='' *explicitly*, which is the same as all interfaces, instead of localhost
+            s = 's' if self.internal_ssl else ''
+            bind_url = f"http{s}://{self.ip}:{self.port}{base_url}"
+            env["JUPYTERHUB_SERVICE_URL"] = bind_url
```

Comment:
This is the remaining question mark, I think: Current behavior is to specify ip or port only *if defined*, leaving it to the subprocess to pick if they are not defined. That means a default of `127.0.0.1:8888` in most circumstances, whereas the defaults on the Spawner are `:0`, which is random port on all interfaces, if passed explicitly. Local spawners explicitly have `if self.port == 0: self.port = random_port()` in `start`, but the *base* Spawner class does not.

However, specifying via URL means it's not possible to specify only ip or port and leave the other as default: both must be specified. Further, `ip=''` has a meaning, which is all interfaces, not 'use the default'. Same for port: `0` means random port, not some default. That means specifying only the port (as was done in LocalProcessSpawner) would result in changing the default behavior from listening on localhost only to all interfaces.

Broad categories of ip/port config patterns:

1. local spawners (localprocess, sudo, systemd) where it makes sense to use `127.0.0.1:0` with `:0` handled in the Spawner, and transformed to an [explicit port](https://github.com/jupyterhub/jupyterhub/blob/574d34388135cceb6c3a633038982529c1b18309/jupyterhub/spawner.py#L1459) prior to launch
2. isolated spawners (docker, k8s, remote VM), where it makes sense to use fixed, explicit `0.0.0.0:8888` (docker + host networking is a wrinkle here, where it should actually do the same as LocalProcessSpawner)
3. subprocess actually does decide (batchspawner, #2727), where it really is reasonable to give the subprocess `:0` and use some special Spawner machinery to report back what host/port were selected and how to connect to it

Possible solutions:

- keep current behavior, where JUPYTERHUB_SERVICE_URL is only defined if one or both of ip and port is specified. The weird behavior: **setting port changes ip from default localhost to all interfaces** which makes me think we should do something else
- make default ip 127.0.0.1 by default, matching current *effective* defaults. This would also mean JUPYTERHUB_SERVICE_URL is always set (by default, at least), which means the singleuser server's internally determined defaults would ~never be used
- always pass JUPYTERHUB_SERVICE_URL, remove case where JUPYTERHUB_SERVICE_URL can be skipped, changing the interpretaion of `ip=''` from "unset" to "specified all interfaces". The default can still be either localhost or all interfaces.


My inclination is to set the default ip, port on the base Spawner class to match the local spawners `127.0.0.1:0` and always pass this as JUPYTERHUB_SERVICE_URL. Then:

- local spawners need to map `if self.port == 0: self.port = random_port()`, which they already do. They do not tend to override default ip, as the existing localhost default works for them.
- container spawners need to specify defaults for both ip, port, which they already do
- if ip is ever specified as `''` this will be explicitly passed and interpreted as 'all interfaces'. This never happens, currently

I think this is the most consistent, backward-compatible approach, where most existing Spawners would need no changes at all, but I'm open to other ideas.

Reply:
I don't have enough knowledge of this at the moment, but I think always setting `JUPYTERHUB_SERVICE_URL` makes sense.